ansible_builder/cli.py

```diff
@@ -134,18 +141,40 @@ def parse_args(args=sys.argv[1:]):
             'This is targeted toward collection authors and maintainers.'
         )
     )
-    add_introspect_options(introspect_parser)
+    introspect_parser.add_argument('--sanitize', action='store_true', help=('Sanitize and de-duplicate requirements. '
+                                                                            'This is normally done separately from the introspect script, but this '
+                                                                            'option is given to more accurately test collection content.'))
+    introspect_parser.add_argument(
+        'folder', default=base_collections_path, nargs='?',
+        help=(
+            'Ansible collections path(s) to introspect. '
+            'This should have a folder named ansible_collections inside of it.'
+        )
+    )
+    # TODO: If we can allow python-builder scripts to be fed multiple files
+    # then we should prefer that over the --user-* options
+    introspect_parser.add_argument(
+        '--user-pip', dest='user_pip',
```

Comment:
@pabelanger sometime in the past, I think you made the point that there's no need to combine `requirements.txt` files and/or `bindep.txt` files from collections with the user files. The scripts like `assemble` and `install-from-bindep` could just pass multiple files to `pip3` or `yum`.

Right now, the logic around [EXTRAS](https://github.com/ansible/python-builder-image/blob/5e8cac7ed3315c7bdc165488670fed27492e2fae/scripts/install-from-bindep#L61) does not seem to be enough to support that. I could be wrong, but I think we would need EXTRAS for assemble/install for both pip/bindep. Right now it looks like its only supported for pip-install.

I don't want to go adding feature requests to python-builder, and without that, these options are needed so that we combine the requirements files without the two-stage build.

Reply:
ZUUL_SIBLIGS is how we handle it in the assemble script, which will then pass the install-from-bindep script the proper files.

https://github.com/ansible/python-builder-image/blob/5e8cac7ed3315c7bdc165488670fed27492e2fae/scripts/assemble#L112

Basically, we can pass in the list of collection directories, then iterate over that to install things.

Reply:
> Basically, we can pass in the list of collection directories

A list of paths under `.zuul-siblings/`, right? And that script puts us in `/tmp/src` before we get there.

So those directories have to be in the absolute location `/tmp/src/.zuul-siblings/`.

To appropriately set the stage, the collections are under `/usr/share/ansible/collections/`, with the subdirs known by `introspect`. The user content is in `/build`, but it may not be there.

But this is ignoring the big showstopper that we have encountered again and again. We could set `ENV ZUUL_SIBLINGS=...`, but how and when can we fill in that `...`? We can't view the contents of the collections until we get past the `ansible-galaxy` layer _in the Containerfile_. This is the fundamental tension of ansible-builder.

Maybe it could be done in a way that looks like this?

https://stackoverflow.com/a/37860420

There are a lot of unanswered questions and I have gone off the deep end speculating.

Reply:
Yes, that is right. Remember, 'python-builder' was initially designed to build python projects in containers. So, from openstack project we expected requirements.txt / bindep.txt files to be in specific places (top level). So we never added any of the logic to allow projects to use custom locations.

What we likely should do, is update assemble a little to learn more about the project, like if a galaxy.yml file is found, then start looking for metadata files to point to the right locations.

I feel we could add the logic, but honestly I think we likely should create a script, that lives in ansible-builder which sets a lot of that logic up. Then invokes assemble.sh under the covers to minimize the Dockerfile size.

Reply:
> What we likely should do, is update assemble a little to learn more about the project, like if a galaxy.yml file is found, then start looking for metadata files to point to the right locations.

There's something here I love, and something I'm hesitant about.

The `requirements.yml` (galaxy requirements file) is not at all useful on its own. We can only do meaningful things with the _installed_ collections, and allow space for things like Automation Hub credentials. This is what ansible-builder does.

In the big picture, I would very much like to combine ansible-builder with python-builder. Taking a very long view, a 3-stage build (with python-builder / ansible-runner image build chain) could possibly satisfy both use cases... but there are valid reasons to think this may never be acceptable.

Within the current context, with the new single-stage, 2 phase build, and a new `ansible-builder` image, I still have some more runway. I could do a skunkworks version of providing ZUUL_SIBLINGS, but it'll have `introspect` doing a lot of ugly file moving operations.

Reply:
@pabelanger trying to follow up here, the big-picture thing we talked about here is not something we can do in the near-term. But, I was worried that the current PR would not be highly compatible with the directions you and I are interested in taking things in. So I put this up:

https://github.com/ansible/ansible-builder/pull/217

The idea is that, one day, we may have a python-builder image that contains the builder "introspect" logic in addition to the scripts. That could be implemented with relative ease after that PR. It would be difficult to ever change to that after the changes here to the 2-phase Containerfile. At least, that's my primary motivation.